### PR TITLE
[NFC] add DiagnosticsTestHelper decl

### PR DESCRIPTION
**Pull request by jsji**
This is one of the many PRs to fix errors with LLVM_ENABLE_WERROR=on. Built by GCC 11.

Fix warning

llvm-project/clang/unittests/Basic/DiagnosticTest.cpp:19:6: error: ‘void clang::DiagnosticsTestHelper(clang::DiagnosticsEngine&)’ has not been declared within ‘clang’ [-Werror]
   19 | void clang::DiagnosticsTestHelper(DiagnosticsEngine &diag) {
      |      ^~~~~
In file included from llvm-project/clang/unittests/Basic/DiagnosticTest.cpp:9:
llvm-project/clang/include/clang/Basic/Diagnostic.h:567:15: note: only here as a ‘friend’
  567 |   friend void DiagnosticsTestHelper(DiagnosticsEngine &);

### CODE REVIEW

**Review by Sirraide — COMMENTED** [review]
It feels a bit weird to put a function that is only used in a test right in the middle of a header, doubly so considering that GCC is the only compiler that doesn’t like it if the only declaration of a function is a friend declaration...

**Comment by cor3ntin** [comment]
GCC is being weird https://compiler-explorer.com/z/3nnxehP6T
I don't know of a better fix but this at least needs a comment

File: clang/include/clang/Basic/Diagnostic.h [hunk 1]
```diff
@@ -1024,6 +1024,9 @@ class DiagnosticsEngine : public RefCountedBase<DiagnosticsEngine> {
   /// @}
 };
 
+/// Declare the friend function for testing to avoid GCC warning.
```
Comment (erichkeane) [inline clang/include/clang/Basic/Diagnostic.h:1027]: couldn't we just put this at the top of the unittest file where this is defined/used?  Just do:

`namespace clang {
<declaration>
}`

It seems silly to do this for a function that is only used in the unittests.

**Review by erichkeane — APPROVED** [review]
Fix whatever clang-format hates about this, and please add a detailed comment explaining why we are doing this, else LGTM.

**Comment by jsji** [comment]
CI failures are irrelvant.

**Pull request merged**